engine/src/main/java/org/pentaho/di/trans/steps/orabulkloader/OraBulkLoader.java

```diff
@@ -0,0 +1,691 @@
+                checkExitVal( exitVal );
+              } else if ( !first ) {
+                throw new KettleException( "Internal error: no sqlldr process running" );
+              }
+            } catch ( Exception ex ) {
+              throw new KettleException( "Error while executing sqlldr", ex );
+            }
+          }
+        }
+        return false;
+      }
+
+      if ( !preview ) {
+        if ( first ) {
+          first = false;
+
+          String recTerm = Const.CR;
+          if ( !Utils.isEmpty( meta.getAltRecordTerm() ) ) {
+            recTerm = substituteRecordTerminator( meta.getAltRecordTerm() );
+          }
+
+          createControlFile( environmentSubstitute( meta.getControlFile() ), r, meta );
+          output = new OraBulkDataOutput( meta, recTerm );
+
+          if ( OraBulkLoaderMeta.METHOD_AUTO_CONCURRENT.equals( meta.getLoadMethod() ) ) {
+            execute( meta, false );
+          }
+          output.open( this, sqlldrProcess );
+        }
+        output.writeLine( getInputRowMeta(), r );
+      }
+      putRow( getInputRowMeta(), r );
+      incrementLinesOutput();
+
+    } catch ( KettleException e ) {
+      logError( BaseMessages.getString( PKG, "OraBulkLoader.Log.ErrorInStep" ) + e.getMessage() );
+      setErrors( 1 );
+      stopAll();
+      setOutputDone(); // signal end to receiver(s)
+      return false;
+    }
+
+    return true;
+  }
+
+  protected void verifyDatabaseConnection() throws KettleException {
+    if ( meta.getDatabaseMeta() == null ) {
+      throw new KettleException( BaseMessages.getString( PKG, "OraBulkLoaderMeta.GetSQL.NoConnectionDefined" ) );
+    }
+  }
+
+  public boolean init( StepMetaInterface smi, StepDataInterface sdi ) {
+    meta = (OraBulkLoaderMeta) smi;
+    data = (OraBulkLoaderData) sdi;
+
+    Trans trans = getTrans();
+    preview = trans.isPreview();
+
+    if ( super.init( smi, sdi ) ) {
+      try {
+        verifyDatabaseConnection();
+      } catch ( KettleException ex ) {
+        logError( ex.getMessage() );
+        return false;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  public void dispose( StepMetaInterface smi, StepDataInterface sdi ) {
+    meta = (OraBulkLoaderMeta) smi;
+    data = (OraBulkLoaderData) sdi;
+
+    super.dispose( smi, sdi );
+
+    // close output stream (may terminate running sqlldr)
+    if ( output != null ) {
+      // Close the output
+      try {
+        output.close();
+      } catch ( IOException e ) {
+        logError( "Error while closing output", e );
+      }
+
+      output = null;
+    }
+    // running sqlldr process must be terminated
+    if ( sqlldrProcess != null ) {
+      try {
+        int exitVal = sqlldrProcess.waitFor();
+        sqlldrProcess = null;
+        logBasic( BaseMessages.getString( PKG, "OraBulkLoader.Log.ExitValueSqlldr", "" + exitVal ) );
+      } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)